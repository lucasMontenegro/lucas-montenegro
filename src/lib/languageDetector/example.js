import React, { useState, Fragment } from "react"
import { Route } from "react-router-dom"
import languageDetector from "lib/languageDetector"
import Div from "lib/utils/react/Div"
import GetValue from "lib/utils/react/GetValue"
import SetValue from "lib/utils/react/SetValue"
import CreateDestroy from "lib/utils/react/CreateDestroy"
function LanguageDetector () {
  const [detectorGet, setDetectorGet] = useState(null)
  const [count, setCount] = useState(0)
  const ready = languageDetector.useReadyState()
  return (
    <Div className="LanguageDetector" color="green">
      <h5>languageDetector</h5>
      <SetValue
        className="init"
        value="[`en`, `es`]"
        onClick={() => languageDetector.init([`en`, `es`])}
      />
      <GetValue
        className="get"
        value={detectorGet}
        onClick={() => {
          setDetectorGet(languageDetector.get())
        }}
      />
      {[`en`, `es`].map(languageCode => (
        <SetValue
          className={`set-${languageCode}`}
          value={languageCode}
          onClick={() => languageDetector.set(languageCode)}
        />
      ))}
      <GetValue
        className="count"
        value={count}
        onClick={() => setCount(count => count + 1)}
      />
      <Div color="purple">
        useReadyState:
        &nbsp;
        <span className="useReadyState">{JSON.stringify(ready)}</span>
      </Div>
    </Div>
  )
}
function Example () {
  const box = 900
  return (
    <Fragment>
      <CreateDestroy id="languageDetector1" box={box} Component={LanguageDetector} />
      <CreateDestroy id="languageDetector2" box={box} Component={LanguageDetector} />
    </Fragment>
  )
}
export default (<Route exact path="/languageDetector" component={Example} />)