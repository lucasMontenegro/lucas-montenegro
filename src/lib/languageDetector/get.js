import i18next from "i18next"
export default function get () {
  return i18next.language
}