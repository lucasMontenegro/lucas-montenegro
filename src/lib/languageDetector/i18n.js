import i18next from "i18next"
import LanguageDetector from "i18next-browser-languagedetector"
import supportedLanguages from "languages/supported"
import isProduction from "lib/utils/isProduction"
export function init () {
  if (i18next.isInitialized) return
  i18next
    .use(LanguageDetector)
    .init({
      debug: !isProduction(),
      fallbackLng: `en`,
      whitelist: supportedLanguages,
    })
}
export function get () {
  return i18next.language
}
export function set (languageCode) {
  i18next.changeLanguage(languageCode)
}