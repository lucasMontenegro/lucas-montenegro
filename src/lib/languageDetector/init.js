import i18next from "i18next"
import LanguageDetector from "i18next-browser-languagedetector"
import isProduction from "lib/utils/isProduction"
export default function init (whitelist) {
  if (i18next.isInitialized) return
  i18next
    .use(LanguageDetector)
    .init({
      debug: !isProduction(),
      fallbackLng: whitelist[0],
      whitelist,
    })
}