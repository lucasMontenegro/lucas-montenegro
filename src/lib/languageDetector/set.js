import i18next from "i18next"
export default function set (languageCode) {
  i18next.changeLanguage(languageCode)
}