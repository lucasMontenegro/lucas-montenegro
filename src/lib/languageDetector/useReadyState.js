import { useLayoutEffect, useState } from "react"
import i18next from "i18next"
export default function useReadyState () {
  const [ready, setReady] = useState(false)
  useLayoutEffect(() => {
    if (i18next.isInitialized) {
      setReady(true)
    } else {
      function handleInit () {
        setReady(true)
      }
      i18next.on(`initialized`, handleInit)
      return () => i18next.off(`initialized`, handleInit)
    }
  }, [])
  return ready
}