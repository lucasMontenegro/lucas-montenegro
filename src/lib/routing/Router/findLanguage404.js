import languageDetector from "lib/languageDetector"
export default function findLanguage404 (location) {
  const route = this.routes.clientNotFound.find(r => r.match(location))
  if (route) {
    const { languageCode } = route
    languageDetector.set(languageCode)
    return {
      render: this.renderEmpty,
      languageCode,
      redirect: {
        ...this.locations.notFound[languageCode],
        state: location,
      },
    }
  }
  return null
}