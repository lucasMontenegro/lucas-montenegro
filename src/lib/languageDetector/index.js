import { useLayoutEffect, useState } from "react"
import i18next from "i18next"
import { init, get, set } from "lib/languageDetector/i18n"
export default {
  init,
  get,
  set,
  useReadyState () {
    const [ready, setReady] = useState(false)
    useLayoutEffect(() => {
      if (i18next.isInitialized) {
        setReady(true)
      } else {
        function handleInit () {
          setReady(true)
        }
        i18next.on(`initialized`, handleInit)
        return () => i18next.off(`initialized`, handleInit)
      }
    }, [])
    return ready
  },
}