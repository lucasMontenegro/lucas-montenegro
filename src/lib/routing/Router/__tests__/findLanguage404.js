import languageDetector from "lib/languageDetector"
import findLanguage404 from "lib/routing/Router/findLanguage404"
jest.mock(`lib/languageDetector`, () => ({
  __esModule: true,
  default: { set: jest.fn() },
}))
describe(`lib/routing/Router/findLanguage404`, () => {
  it(`should use the right dependency versions`, () => {
    expect(jestUtils.getDependencies([`react-router-dom`])).toMatchSnapshot()
  })
  let match
  const router = {
    findLanguage404,
    routes: {
      clientNotFound: [
        { languageCode: `foo`, match: () => false },
        { languageCode: `bar`, match: () => match },
        { languageCode: `baz`, match: () => false },
      ],
    },
    renderEmpty: {},
    locations: {
      notFound: { bar: { pathname: `bar` } },
    },
  }
  describe(`findLanguage404 (routes.clientNotFound.find => null)`, () => {
    let result
    beforeAll(() => {
      match = false
      result = router.findLanguage404({})
    })
    it(`should not call languageDetector.set`, () => {
      expect(languageDetector.set.mock.calls).toEqual([])
    })
    it(`should return null`, () => {
      expect(result).toBeNull()
    })
  })
  describe(`findLanguage404 (routes.clientNotFound.find => route)`, () => {
    let result
    const location = {}
    beforeAll(() => {
      match = true
      result = router.findLanguage404(location)
    })
    it(`should call languageDetector.set`, () => {
      expect(languageDetector.set.mock.calls).toEqual([[`bar`]])
      languageDetector.set.mockClear()
    })
    it(`should return a route that redirects to notFound`, () => {
      expect(result).toEqual({
        render: {},
        languageCode: `bar`,
        redirect: { pathname: `bar`, state: {} },
      })
      expect(result.render).toBe(router.renderEmpty)
      expect(result.redirect.state).toBe(location)
    })
  })
})