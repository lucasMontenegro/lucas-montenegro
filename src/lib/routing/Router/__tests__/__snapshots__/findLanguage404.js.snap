// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`lib/routing/Router/findLanguage404 should use the right dependency versions 1`] = `
Object {
  "react-router-dom": "^5.1.2",
}
`;
